core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -600,10 +611,14 @@ public OutputSource createOutputSource(Set<TaskId> selectedTasks)
             verify(exchange != null, "exchange is expected to be initialized");
             verify(sinkInstanceHandle != null, "sinkInstanceHandle is expected to be initialized");
 
-            exchangeSink.finish();
```

Comment:
Basically the `ExchangeSink#finish` operation and the `Exchange#getSourceHandles` operation both can take a substantial amount of time. But since they are both non blocking the `ExchangeClient` thread doesn't have to be blocked for all this time.